File: packages/server/src/services/stack-detector.service.ts
import { Octokit } from 'octokit';
import { join } from 'path';
import { createLogger } from '../utils/logger.js';
import { getErrorMessage } from '../utils/errors.js';

/**
 * Detected technology stack
 */
export interface DetectedStack {
  framework: string | null;
  state_management: string | null;
  styling: string | null;
  testing: string | null;
}

/**
 * Response from stack detection
 */
export interface StackDetectionResponse {
  detected_stack: DetectedStack;
  confidence: {
    framework: number;
    state_management: number;
    styling: number;
    testing: number;
  };
}

/**
 * Default empty detected stack
 */
export const DEFAULT_DETECTED_STACK: DetectedStack = {
  framework: null,
  state_management: null,
  styling: null,
  testing: null,
};

const logger = createLogger('stack-detector');

/**
 * Configuration for common frameworks and libraries to detect
 */
const FRAMEWORK_PATTERNS: Record<string, string[]> = {
  'Next.js': ['next', '@next/'],
  'React': ['react', 'react-dom'],
  'Vue.js': ['vue', '@vue/'],
  'Angular': ['@angular/core', '@angular/'],
  'Svelte': ['svelte', '@sveltejs/'],
  'Nuxt.js': ['nuxt', '@nuxt/'],
  'Remix': ['@remix-run/'],
  'Astro': ['astro'],
  'Express': ['express'],
  'NestJS': ['@nestjs/core', '@nestjs/'],
  'Fastify': ['fastify'],
  'Hono': ['hono'],
};

const STATE_MANAGEMENT_PATTERNS: Record<string, string[]> = {
  'Zustand': ['zustand'],
  'Redux': ['redux', '@reduxjs/toolkit', 'react-redux'],
  'Jotai': ['jotai'],
  'Recoil': ['recoil'],
  'MobX': ['mobx', 'mobx-react'],
  'Pinia': ['pinia'],
  'Vuex': ['vuex'],
  'TanStack Query': ['@tanstack/react-query', 'react-query'],
  'SWR': ['swr'],
  'XState': ['xstate'],
};

const STYLING_PATTERNS: Record<string, string[]> = {
  'Tailwind CSS': ['tailwindcss', '@tailwindcss/'],
  'styled-components': ['styled-components'],
  'Emotion': ['@emotion/react', '@emotion/styled'],
  'CSS Modules': [], // Detected by file patterns
  'Sass': ['sass', 'node-sass'],
  'Less': ['less'],
  'Chakra UI': ['@chakra-ui/react'],
  'Material UI': ['@mui/material', '@material-ui/'],
  'Ant Design': ['antd'],
  'shadcn/ui': [], // Detected by file patterns
};

const TESTING_PATTERNS: Record<string, string[]> = {
  'Vitest': ['vitest'],
  'Jest': ['jest', '@jest/'],
  'Playwright': ['@playwright/test', 'playwright'],
  'Cypress': ['cypress'],
  'Testing Library': ['@testing-library/'],
  'Mocha': ['mocha'],
  'AVA': ['ava'],
};

interface PackageJson {
  dependencies?: Record<string, string>;
  devDependencies?: Record<string, string>;
}

interface FileInfo {
  name: string;
  path: string;
  type: 'file' | 'dir';
}

/**
 * Service for detecting the technology stack of a repository
 */
export class StackDetectorService {
  private octokit: Octokit;

  constructor(githubToken: string) {
    this.octokit = new Octokit({ auth: githubToken });
  }

  /**
   * Detect the technology stack of a repository
   */
  async detectStack(owner: string, repo: string, branch: string = 'main'): Promise<StackDetectionResponse> {
    logger.info('Detecting stack for repository', { owner, repo, branch });

    try {
      // Get package.json content
      const packageJson = await this.getPackageJson(owner, repo, branch);

      // Get root files for additional detection
      const rootFiles = await this.getRootFiles(owner, repo, branch);

      // Detect each category
      const [framework, frameworkConfidence] = this.detectCategory(packageJson, FRAMEWORK_PATTERNS);
      const [stateManagement, stateConfidence] = this.detectCategory(packageJson, STATE_MANAGEMENT_PATTERNS);
      let [styling, stylingConfidence] = this.detectCategory(packageJson, STYLING_PATTERNS);
      const [testing, testingConfidence] = this.detectCategory(packageJson, TESTING_PATTERNS);

      // Additional file-based detection
      if (!styling && rootFiles.some(f => f.name === 'tailwind.config.js' || f.name === 'tailwind.config.ts')) {
        styling = 'Tailwind CSS';
        stylingConfidence = 0.9;
      }

      // Check for shadcn/ui by looking for components.json
      if (rootFiles.some(f => f.name === 'components.json')) {
        if (styling) {
          styling = `${styling}, shadcn/ui`;
        } else {
          styling = 'shadcn/ui';
          stylingConfidence = 0.9;
        }
      }

      const detectedStack: DetectedStack = {
        framework,
        state_management: stateManagement,
        styling,
        testing,
      };

      const result: StackDetectionResponse = {
        detected_stack: detectedStack,
        confidence: {
          framework: frameworkConfidence,
          state_management: stateConfidence,
          styling: stylingConfidence,
          testing: testingConfidence,
        },
      };

      logger.info('Stack detected successfully', {
        owner,
        repo,
        stack: detectedStack,
      });

      return result;
    } catch (error) {
      const errorMessage = getErrorMessage(error);
      logger.error('Failed to detect stack', { owner, repo, error: errorMessage });

      // Return empty stack on error
      return {
        detected_stack: {
          framework: null,
          state_management: null,
          styling: null,
          testing: null,
        },
        confidence: {
          framework: 0,
          state_management: 0,
          styling: 0,
          testing: 0,
        },
      };
    }
  }

  /**
   * Get package.json from repository
   */
  private async getPackageJson(owner: string, repo: string, branch: string): Promise<PackageJson | null> {
    try {
      const response = await this.octokit.rest.repos.getContent({
        owner,
        repo,
        path: 'package.json',
        ref: branch,
      });

      if ('content' in response.data) {
        const content = Buffer.from(response.data.content, 'base64').toString('utf-8');
        return JSON.parse(content) as PackageJson;
      }
    } catch (error) {
      logger.debug('Could not fetch package.json', { owner, repo, error: getErrorMessage(error) });
    }
    return null;
  }

  /**
   * Get list of files in repository root
   */
  private async getRootFiles(owner: string, repo: string, branch: string): Promise<FileInfo[]> {
    try {
      const response = await this.octokit.rest.repos.getContent({
        owner,
        repo,
        path: '',
        ref: branch,
      });

      if (Array.isArray(response.data)) {
        return response.data.map(item => ({
          name: item.name,
          path: item.path,
          type: item.type as 'file' | 'dir',
        }));
      }
    } catch (error) {
      logger.debug('Could not fetch root files', { owner, repo, error: getErrorMessage(error) });
    }
    return [];
  }

  /**
   * Detect a category (framework, state management, etc.) from package.json
   */
  private detectCategory(
    packageJson: PackageJson | null,
    patterns: Record<string, string[]>
  ): [string | null, number] {
    if (!packageJson) {
      return [null, 0];
    }

    const allDeps = {
      ...packageJson.dependencies,
      ...packageJson.devDependencies,
    };

    const depNames = Object.keys(allDeps);

    for (const [name, searchPatterns] of Object.entries(patterns)) {
      if (searchPatterns.length === 0) continue;

      for (const pattern of searchPatterns) {
        const found = depNames.some(dep =>
          dep === pattern || dep.startsWith(pattern)
        );
        if (found) {
          // Higher confidence if it's a direct match
          const confidence = depNames.includes(pattern) ? 1.0 : 0.9;
          return [name, confidence];
        }
      }
    }

    return [null, 0];
  }
}

/**
 * Create a stack detector service with the given GitHub token
 */
export function createStackDetector(githubToken: string): StackDetectorService {
  return new StackDetectorService(githubToken);
}

/**
 * Service for detecting the technology stack of a LOCAL repository (filesystem-based, no GitHub API)
 */
export class LocalStackDetectorService {
  /**
   * Detect the technology stack from a local repository path
   */
  async detectStack(repoPath: string): Promise<StackDetectionResponse> {
    logger.info('Detecting stack for local repository', { repoPath });

    try {
      const packageJson = await this.getPackageJson(repoPath);
      const rootFiles = await this.getRootFiles(repoPath);

      // Reuse the same detection logic pattern
      const [framework, frameworkConfidence] = this.detectCategory(packageJson, FRAMEWORK_PATTERNS);
      const [stateManagement, stateConfidence] = this.detectCategory(packageJson, STATE_MANAGEMENT_PATTERNS);
      let [styling, stylingConfidence] = this.detectCategory(packageJson, STYLING_PATTERNS);
      const [testing, testingConfidence] = this.detectCategory(packageJson, TESTING_PATTERNS);

      // Additional file-based detection
      if (!styling && rootFiles.some(f => f === 'tailwind.config.js' || f === 'tailwind.config.ts')) {
        styling = 'Tailwind CSS';
        stylingConfidence = 0.9;
      }

      if (rootFiles.some(f => f === 'components.json')) {
        if (styling) {
          styling = `${styling}, shadcn/ui`;
        } else {
          styling = 'shadcn/ui';
          stylingConfidence = 0.9;
        }
      }

      const detectedStack: DetectedStack = {
        framework,
        state_management: stateManagement,
        styling,
        testing,
      };

      return {
        detected_stack: detectedStack,
        confidence: {
          framework: frameworkConfidence,
          state_management: stateConfidence,
          styling: stylingConfidence,
          testing: testingConfidence,
        },
      };
    } catch (error) {
      const errorMessage = getErrorMessage(error);
      logger.error('Failed to detect local stack', { repoPath, error: errorMessage });
      return {
        detected_stack: DEFAULT_DETECTED_STACK,
        confidence: { framework: 0, state_management: 0, styling: 0, testing: 0 },
      };
    }
  }

  private async getPackageJson(repoPath: string): Promise<PackageJson | null> {
    try {
      const { readFile } = await import('fs/promises');
      const content = await readFile(join(repoPath, 'package.json'), 'utf-8');
      return JSON.parse(content) as PackageJson;
    } catch {
      return null;
    }
  }

  private async getRootFiles(repoPath: string): Promise<string[]> {
    try {
      const { readdir } = await import('fs/promises');
      const entries = await readdir(repoPath);
      return entries;
    } catch {
      return [];
    }
  }

  private detectCategory(
    packageJson: PackageJson | null,
    patterns: Record<string, string[]>
  ): [string | null, number] {
    if (!packageJson) return [null, 0];

    const allDeps = {
      ...packageJson.dependencies,
      ...packageJson.devDependencies,
    };
    const depNames = Object.keys(allDeps);

    for (const [name, searchPatterns] of Object.entries(patterns)) {
      if (searchPatterns.length === 0) continue;
      for (const pattern of searchPatterns) {
        const found = depNames.some(dep => dep === pattern || dep.startsWith(pattern));
        if (found) {
          const confidence = depNames.includes(pattern) ? 1.0 : 0.9;
          return [name, confidence];
        }
      }
    }
    return [null, 0];
  }
}

export function createLocalStackDetector(): LocalStackDetectorService {
  return new LocalStackDetectorService();
}

File: packages/server-rs/src/services/stack_detector_service.rs
//! Stack detector service for analyzing a repository to detect its technology stack.
//!
//! Port of `packages/server/src/services/stack-detector.service.ts`.
//!
//! Detects:
//! - Framework (React, Next.js, Vue, Angular, Svelte, Express, NestJS, etc.)
//! - State management (Zustand, Redux, Jotai, etc.)
//! - Styling (Tailwind, styled-components, Chakra UI, etc.)
//! - Testing (Vitest, Jest, Playwright, Cypress, etc.)
//!
//! Works on local filesystem paths (no GitHub API required).

use std::collections::HashMap;
use std::path::Path;

use serde::{Deserialize, Serialize};
use tokio::fs;
use tracing::info;

use crate::error::AppError;
use crate::models::repository::DetectedStack;

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

/// Response from stack detection, including confidence scores.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct StackDetectionResponse {
    pub detected_stack: DetectedStack,
    pub confidence: StackConfidence,
}

/// Confidence scores for each detected category.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct StackConfidence {
    pub framework: f64,
    pub state_management: f64,
    pub styling: f64,
    pub testing: f64,
}

/// Simplified package.json representation for dependency scanning.
#[derive(Debug, Deserialize)]
struct PackageJson {
    #[serde(default)]
    dependencies: HashMap<String, String>,
    #[serde(rename = "devDependencies", default)]
    dev_dependencies: HashMap<String, String>,
}

// ---------------------------------------------------------------------------
// Detection patterns (mirrors the TS constants)
// ---------------------------------------------------------------------------

fn framework_patterns() -> Vec<(&'static str, Vec<&'static str>)> {
    vec![
        ("Next.js", vec!["next", "@next/"]),
        ("React", vec!["react", "react-dom"]),
        ("Vue.js", vec!["vue", "@vue/"]),
        ("Angular", vec!["@angular/core", "@angular/"]),
        ("Svelte", vec!["svelte", "@sveltejs/"]),
        ("Nuxt.js", vec!["nuxt", "@nuxt/"]),
        ("Remix", vec!["@remix-run/"]),
        ("Astro", vec!["astro"]),
        ("Express", vec!["express"]),
        ("NestJS", vec!["@nestjs/core", "@nestjs/"]),
        ("Fastify", vec!["fastify"]),
        ("Hono", vec!["hono"]),
    ]
}

fn state_management_patterns() -> Vec<(&'static str, Vec<&'static str>)> {
    vec![
        ("Zustand", vec!["zustand"]),
        ("Redux", vec!["redux", "@reduxjs/toolkit", "react-redux"]),
        ("Jotai", vec!["jotai"]),
        ("Recoil", vec!["recoil"]),
        ("MobX", vec!["mobx", "mobx-react"]),
        ("Pinia", vec!["pinia"]),
        ("Vuex", vec!["vuex"]),
        ("TanStack Query", vec!["@tanstack/react-query", "react-query"]),
        ("SWR", vec!["swr"]),
        ("XState", vec!["xstate"]),
    ]
}

fn styling_patterns() -> Vec<(&'static str, Vec<&'static str>)> {
    vec![
        ("Tailwind CSS", vec!["tailwindcss", "@tailwindcss/"]),
        ("styled-components", vec!["styled-components"]),
        ("Emotion", vec!["@emotion/react", "@emotion/styled"]),
        ("Sass", vec!["sass", "node-sass"]),
        ("Less", vec!["less"]),
        ("Chakra UI", vec!["@chakra-ui/react"]),
        ("Material UI", vec!["@mui/material", "@material-ui/"]),
        ("Ant Design", vec!["antd"]),
    ]
}

fn testing_patterns() -> Vec<(&'static str, Vec<&'static str>)> {
    vec![
        ("Vitest", vec!["vitest"]),
        ("Jest", vec!["jest", "@jest/"]),
        ("Playwright", vec!["@playwright/test", "playwright"]),
        ("Cypress", vec!["cypress"]),
        ("Testing Library", vec!["@testing-library/"]),
        ("Mocha", vec!["mocha"]),
        ("AVA", vec!["ava"]),
    ]
}

// ---------------------------------------------------------------------------
// Detection logic
// ---------------------------------------------------------------------------

/// Detects a single category (framework, styling, etc.) from the package.json deps.
fn detect_category(
    dep_names: &[String],
    patterns: &[(&str, Vec<&str>)],
) -> (Option<String>, f64) {
    for (name, search_patterns) in patterns {
        if search_patterns.is_empty() {
            continue;
        }
        for pattern in search_patterns {
            let found = dep_names
                .iter()
                .any(|dep| dep == pattern || dep.starts_with(pattern));
            if found {
                let confidence = if dep_names.iter().any(|d| d == pattern) {
                    1.0
                } else {
                    0.9
                };
                return (Some(name.to_string()), confidence);
            }
        }
    }
    (None, 0.0)
}

// ---------------------------------------------------------------------------
// Public API
// ---------------------------------------------------------------------------

/// Detects the technology stack of a local repository by reading `package.json`,
/// `Cargo.toml`, `go.mod`, etc. from the filesystem.
pub async fn detect_stack(repo_path: &Path) -> Result<StackDetectionResponse, AppError> {
    info!(repo_path = %repo_path.display(), "Detecting stack for local repository");

    let package_json = read_package_json(repo_path).await;
    let root_files = list_root_files(repo_path).await;

    // Detect from package.json dependencies
    let dep_names = match &package_json {
        Some(pj) => {
            let mut names: Vec<String> = pj.dependencies.keys().cloned().collect();
            names.extend(pj.dev_dependencies.keys().cloned());
            names
        }
        None => Vec::new(),
    };

    let (mut framework, framework_conf) = detect_category(&dep_names, &framework_patterns());
    let (state_mgmt, state_conf) = detect_category(&dep_names, &state_management_patterns());
    let (mut styling, mut styling_conf) = detect_category(&dep_names, &styling_patterns());
    let (testing, testing_conf) = detect_category(&dep_names, &testing_patterns());

    // File-based detection for styling
    if styling.is_none()
        && root_files
            .iter()
            .any(|f| f == "tailwind.config.js" || f == "tailwind.config.ts")
    {
        styling = Some("Tailwind CSS".to_string());
        styling_conf = 0.9;
    }

    // shadcn/ui detection via components.json
    if root_files.iter().any(|f| f == "components.json") {
        if let Some(ref existing) = styling {
            styling = Some(format!("{existing}, shadcn/ui"));
        } else {
            styling = Some("shadcn/ui".to_string());
            styling_conf = 0.9;
        }
    }

    // If no JS framework detected, try other project types
    if framework.is_none() {
        if root_files.iter().any(|f| f == "Cargo.toml") {
            framework = Some("Rust".to_string());
        } else if root_files.iter().any(|f| f == "go.mod") {
            framework = Some("Go".to_string());
        } else if root_files
            .iter()
            .any(|f| f == "requirements.txt" || f == "pyproject.toml")
        {
            framework = Some("Python".to_string());
        } else if root_files.iter().any(|f| f == "Gemfile") {
            framework = Some("Ruby".to_string());
        } else if root_files.iter().any(|f| f == "pom.xml" || f == "build.gradle") {
            framework = Some("Java".to_string());
        }
    }

    let detected_stack = DetectedStack {
        framework,
        state_management: state_mgmt,
        styling,
        testing,
    };

    info!(
        repo_path = %repo_path.display(),
        ?detected_stack,
        "Stack detected successfully"
    );

    Ok(StackDetectionResponse {
        detected_stack,
        confidence: StackConfidence {
            framework: framework_conf,
            state_management: state_conf,
            styling: styling_conf,
            testing: testing_conf,
        },
    })
}

/// Reads `package.json` from a directory, returning `None` on any error.
async fn read_package_json(dir: &Path) -> Option<PackageJson> {
    let path = dir.join("package.json");
    let content = fs::read_to_string(&path).await.ok()?;
    serde_json::from_str(&content).ok()
}

/// Lists the names of files and directories in the root of the given path.
async fn list_root_files(dir: &Path) -> Vec<String> {
    let mut names = Vec::new();
    let mut entries = match fs::read_dir(dir).await {
        Ok(e) => e,
        Err(_) => return names,
    };
    while let Ok(Some(entry)) = entries.next_entry().await {
        if let Some(name) = entry.file_name().to_str() {
            names.push(name.to_string());
        }
    }
    names
}

/// Detects the primary language of a repository based on root files.
/// Used by the local scan service.
pub async fn detect_language(repo_path: &Path, has_package_json: bool) -> Option<String> {
    if has_package_json {
        let tsconfig = repo_path.join("tsconfig.json");
        if fs::metadata(&tsconfig).await.is_ok() {
            return Some("TypeScript".to_string());
        }
        return Some("JavaScript".to_string());
    }

    let language_files: &[(&str, &str)] = &[
        ("Cargo.toml", "Rust"),
        ("go.mod", "Go"),
        ("requirements.txt", "Python"),
        ("pyproject.toml", "Python"),
        ("Gemfile", "Ruby"),
        ("pom.xml", "Java"),
        ("build.gradle", "Java"),
        ("composer.json", "PHP"),
    ];

    for (file, lang) in language_files {
        if fs::metadata(repo_path.join(file)).await.is_ok() {
            return Some(lang.to_string());
        }
    }

    None
}
